src_c/mouse.c

```diff
@@ -309,73 +325,119 @@ mouse_set_cursor(PyObject *self, PyObject *args)
+    SDL_Surface *surf = NULL;
+    surf = pgSurface_AsSurface(surfobj);
+
+    cursor = SDL_CreateColorCursor(surf, spotx, spoty);
+    if (!cursor)
         return RAISE(pgExc_SDLError, SDL_GetError());
-    }
+
     lastcursor = SDL_GetCursor();
     SDL_SetCursor(cursor);
     SDL_FreeCursor(lastcursor);
-#endif
+
+    cursor_data.type = 2;
+    cursor_data.spotx = spotx;
+    cursor_data.spoty = spoty;
+    cursor_data.surfobj = surfobj;
     Py_RETURN_NONE;
-}
 
+#endif
+return RAISE(PyExc_TypeError, "Cursors from a surface are unavailable in SDL1");
+}
 
-#if IS_SDLv2
 static PyObject *
-mouse_get_cursor(PyObject *self)
+mouse_set_system_cursor(PyObject *self, PyObject *args)
 {
-    return RAISE(PyExc_TypeError, "The get_cursor method is unavailable in SDL2");
+    int constant;
+    
+    VIDEO_INIT_CHECK();
+
+    if (!PyArg_ParseTuple(args, "i", &constant)) {
+        return NULL;
+    }
+
+    return _set_system_cursor(constant);
 }
-#else /* IS_SDLv1*/
+
+//this function isn't all of mouse.set_cursor, it goes through a python function first
 static PyObject *
-mouse_get_cursor(PyObject *self)
+mouse_set_cursor(PyObject *self, PyObject *args, PyObject *kwds)
 {
-    SDL_Cursor *cursor = NULL;
-    PyObject *xordata, *anddata;
-    int size, loop, w, h, spotx, spoty;
+    //normal_cursor stuff
+    int w=0, h=0, spotx, spoty;
+    PyObject *xormask, *andmask;
 
-    VIDEO_INIT_CHECK();
+    //system_cursor stuff
+    int constant = -1;
 
-    cursor = SDL_GetCursor();
-    if (!cursor)
-        return RAISE(pgExc_SDLError, SDL_GetError());
+    //color_cursor stuff
+    pgSurfaceObject *surfobj = NULL;
 
-    w = cursor->area.w;
-    h = cursor->area.h;
-    spotx = cursor->hot_x;
-    spoty = cursor->hot_y;
+    static char *keywords[] = {"system", "bitmap", "color", NULL};
 
-    size = cursor->area.w * cursor->area.h / 8;
-    xordata = PyTuple_New(size);
-    if (!xordata)
-        return NULL;
-    anddata = PyTuple_New(size);
-    if (!anddata) {
-        Py_DECREF(xordata);
+    VIDEO_INIT_CHECK();
+
+    if (!PyArg_ParseTupleAndKeywords(args, kwds, "|(i)((ii)(ii)OO)((ii)O!)", keywords, &constant,
+                                     &w, &h, &spotx, &spoty, &xormask, &andmask, &spotx, &spoty,
+                                     &pgSurface_Type, &surfobj)) {
         return NULL;
     }
 
-    for (loop = 0; loop < size; ++loop) {
-        PyTuple_SET_ITEM(xordata, loop, PyInt_FromLong(cursor->data[loop]));
-        PyTuple_SET_ITEM(anddata, loop, PyInt_FromLong(cursor->mask[loop]));
+    if (constant > -1) {
+        return _set_system_cursor(constant);
     }
+    else if (w && h) {
+        return _set_bitmap_cursor(w, h, spotx, spoty, xormask, andmask);
+    }
+    else if (surfobj) {
+        return _set_color_cursor(spotx, spoty, surfobj);
+    }
+    return RAISE(PyExc_ValueError, "Invalid cursor format: no valid template found");  
+}
+
+//should the buildvalue 'O's be replaced by 'N's?
```

Comment:
What is this about?

Reply:
It's another reference count thing.

Objects can be returned with Py_BuildValue with the format "O" or "N". [https://docs.python.org/3/c-api/arg.html#c.Py_BuildValue](https://docs.python.org/3/c-api/arg.html#c.Py_BuildValue)

Using "O" increases the reference count by 1, using "N" doesn't.
But the docs are technical enough I don't really understand the use cases that they suggest.

Reply:
I tested it with "N" and it had intermittent segfaults, so looks like "O" it is.